### PR TITLE
WIP: Improve the speed of searchsorted.

### PR DESCRIPTION
**Pull request by rainwoodman**
Improving the speed of searchsorted.

By peeking the current item in the haystack first we roughly recover a O(n + m) algorithm instead of O(n log m) when the haystack is sorted. For small n and m it is slower.

This PR needs to be cleaned up, and improved. There probably are better heuristics to decide when to use O(n +m) and when to use O(n logm).

Inputs are welcome.

Benchmarks:

I am having difficulty running asv on master due to the compilation error in https://github.com/numpy/numpy/issues/11853

so only a simple comparison for now.

After:
```
Previous HEAD position was eca9936d1... Add benchmark for Search.
Switched to branch 'fastersearch'
(base) [yfeng1@waterfall numpy]$ !py
python runtests.py --bench Search
Building, see build.log...
Build OK
· Discovering benchmarks
· Running 1 total benchmarks (1 commits * 1 environments * 1 benchmarks)
[  0.00%] ·· Building for existing-py_home_yfeng1_anaconda3_install_bin_python
[  0.00%] ·· Benchmarking existing-py_home_yfeng1_anaconda3_install_bin_python
[100.00%] ··· Running bench_function_base.Search.time_1                                                                               ok
[100.00%] ···· 
               ============ ============== ================= ================= ================= =================
               --                                                 side / needletype / dtype                       
               --------------------------- -----------------------------------------------------------------------
                needlesize   haystacksize   l / sorted / f8   l / random / f8   r / sorted / f8   r / random / f8 
               ============ ============== ================= ================= ================= =================
                   1000          1000           14.17μs           49.62μs           12.79μs           48.18μs     
                   1000         10000           21.90μs           81.12μs           20.60μs           79.99μs     
                   1000         100000          82.65μs           195.03μs          82.83μs           192.62μs    
                   1000        1000000           1.35ms            1.51ms            1.34ms            1.52ms     
                  10000          1000           41.77μs           596.70μs          38.65μs           584.59μs    
                  10000         10000           152.83μs          819.17μs          140.88μs          835.26μs    
                  10000         100000          224.80μs           1.50ms           216.97μs           1.49ms     
                  10000        1000000           2.02ms            4.08ms            2.00ms            4.15ms     
                  100000         1000           219.59μs           5.79ms           220.17μs           5.71ms     
                  100000        10000           436.52μs           8.18ms           391.23μs           8.07ms     
                  100000        100000           1.82ms           14.47ms            1.63ms           14.30ms     
                  100000       1000000           3.45ms           30.56ms            3.33ms           30.12ms     
                 1000000         1000            2.05ms           58.13ms            2.05ms           57.47ms     
                 1000000        10000            2.28ms           81.98ms            2.27ms           82.95ms     
                 1000000        100000           4.71ms           145.07ms           4.58ms           143.52ms    
                 1000000       1000000          22.75ms           266.39ms          19.95ms           267.09ms    
               ============ ============== ================= ================= ================= =================
```
before
```
HEAD is now at eca9936d1... Add benchmark for Search.
(base) [yfeng1@waterfall numpy]$ !p
python runtests.py --bench Search
Building, see build.log...
Build OK
· Discovering benchmarks
· Running 1 total benchmarks (1 commits * 1 environments * 1 benchmarks)
[  0.00%] ·· Building for existing-py_home_yfeng1_anaconda3_install_bin_python
[  0.00%] ·· Benchmarking existing-py_home_yfeng1_anaconda3_install_bin_python
[100.00%] ··· Running bench_function_base.Search.time_1                                                                               ok
[100.00%] ···· 
               ============ ============== ================= ================= ================= =================
               --                                                 side / needletype / dtype                       
               --------------------------- -----------------------------------------------------------------------
                needlesize   haystacksize   l / sorted / f8   l / random / f8   r / sorted / f8   r / random / f8 
               ============ ============== ================= ================= ================= =================
                   1000          1000           14.40μs           55.83μs           14.14μs           51.88μs     
                   1000         10000           22.96μs           85.93μs           23.58μs           83.82μs     
                   1000         100000          89.63μs           195.00μs          90.94μs           193.37μs    
                   1000        1000000           1.24ms            1.40ms            1.24ms            1.40ms     
                  10000          1000           125.83μs          585.63μs          124.70μs          574.02μs    
                  10000         10000           161.67μs          822.46μs          163.26μs          816.22μs    
                  10000         100000          249.13μs           1.49ms           243.95μs           1.47ms     
                  10000        1000000           2.08ms            4.25ms            1.97ms            4.22ms     
                  100000         1000            1.13ms            5.83ms            1.15ms            5.72ms     
                  100000        10000            1.61ms            8.15ms            1.61ms            8.08ms     
                  100000        100000           1.96ms           14.29ms            1.96ms           14.34ms     
                  100000       1000000           3.77ms           29.60ms            3.64ms           29.72ms     
                 1000000         1000           11.11ms           58.44ms           11.27ms           57.28ms     
                 1000000        10000           15.09ms           81.65ms           15.39ms           80.74ms     
                 1000000        100000          20.13ms           144.49ms          19.97ms           143.42ms    
                 1000000       1000000          24.42ms           264.65ms          24.61ms           263.74ms    
               ============ ============== ================= ================= ================= =================


```

### CODE REVIEW

File: benchmarks/asv.conf.json [hunk 1]
```diff
@@ -35,7 +35,7 @@
 
     // The Pythons you'd like to test against.  If not provided, defaults
     // to the current version of Python used to run `asv`.
-    "pythons": ["2.7"],
+    "pythons": [ "3.6" ],
```
Comment (charris) [inline benchmarks/asv.conf.json]: Hmm... A change like this probably belongs in a separate PR. @pv Does this impact you?

**Comment by shoyer** [comment]
I think this is missing your changes -- I only see a few lines moved in `binsearch.c.src`. If you edited `binsearch.c` instead, note that that file is autogenerated.

Also, you should remove `site.cfg`.

File: numpy/core/src/npysort/binsearch.c.src [hunk 2]
```diff
@@ -66,16 +66,16 @@ binsearch_@side@_@suff@(const char *arr, const char *key, char *ret,
         }
 
         last_key_val = key_val;
-
+        npy_intp mid_idx = min_idx;
```
Comment (charris) [inline numpy/core/src/npysort/binsearch.c.src]: No declarations after statements until we go to C99.

**Comment by charris** [comment]
The `site.cfg` file doesn't belong here, you need to be more careful about commits.

File: numpy/core/src/npysort/binsearch.c.src (hunk 2 above)
Comment (charris) [reply numpy/core/src/npysort/binsearch.c.src]: This will fail for 2.7 on windows, see azure-pipeline test results.

**Comment by charris** [comment]
I'm going to close this. @rainwoodman When you get things fixed up, make another PR.

**Comment by rainwoodman** [comment]
Oops I was hoping to get a few comments about the potential effect on the performance; but most were about the PR itselfs. The commits are not cleaned up because this was a WIP -- I was trying to find a way to workaround the header include order bug and get asv working.

It will slightly slow some cases while drastically increase the speed for others. How acceptable is this?

**Comment by rainwoodman** [comment]
> I think this is missing your changes -- I only see a few lines moved in `binsearch.c.src`. If you edited `binsearch.c` instead, note that that file is autogenerated.
> 
> Also, you should remove `site.cfg`.

The change was that small!

**Comment by shoyer** [comment]
Feel free to reopen after removing the `site.cfg` file

**Comment by shoyer** [comment]
> The change was that small!

OK, I see now this works now. Very simple, I like it! I'll ponder more on the performance implications.

**Comment by shoyer** [comment]
Could you possibly make some lines plots with the most relevant performance comparisons? It's hard to understand a big table of numbers.

**Comment by charris** [comment]
I think `likely` can also help in some cases. @juliantaylor Thoughts?

**Comment by rainwoodman** [comment]
Some progress on this:

http://rainwoodman.github.io/numpy/#bench_function_base.Search.time_sorted?branch=fastersearch-omn&branch=fastersearch-baseline&branch=fastersearch-checkfirst&x-axis=needlesize&p-side='l'&p-haystacksize=10000

The O(m+n) algorithm has a significant advantage when the length of the sorted segment in needles array is longer or comparable to a number that is related to the length of the haystack. I think we can find the length of the sorted segment and decide which algorithm to use on the fly.
